OPHD/Common.h

```diff
@@ -366,3 +366,8 @@ const auto formatDiff = [](int diff)
 {
 	return ((diff > 0) ? "+" : "") + std::to_string(diff);
 };
+
+/**
+ * Silence compiler warnings about unused variable
+ */
+#define UNUSED(x) (void)(x)
```

Comment:
This makes me a bit squeamish. I like to avoid macros whenever possible. Sometimes you can avoid unused variable warnings by commenting out the name of function parameters. For structured bindings, well, you could rewrite to avoid using structured bindings.

Reply:
what's wrong with macros? They're a powerful tool. Rewriting code to not use structured bindings just because of a false positive compiler warning seems odd.

Reply:
It's another stylistic choice -- we tend to avoid macros because they're a great source of errors and abuses of type system issues. Instead we've been using C++17 constexpr and similar constructs in place of the preprocessor.

Reply:
I highly recommend doing a Google search for why C++ macros shouldn't be used.

I think the main one for me is they don't respect scope. This can lead to lots of nasty conflicts, particularly with the Windows headers which define a lot of macros. If you `#define max` as a macro function, then try to use `std::max`, the macro will do a textual replacement on the call to `std::max` and you'll end up with something weird like `std::(a<b?a:b)`. Have fun diagnosing the syntax error on that, particularly since you don't normally see the result of pre-processing and may have no idea a macro was ever even invoked there.

There's issues about multiple evaluation of parameters when passed expressions with side effects. How many times are `x` and `y` incremented below? Maybe once? Maybe twice? Maybe it depends on how they compare?
```cpp
#define max(a, b) a < b ? a : b
max(x++, y++); // x++ < y++ ? x++ : y++
```

They also have issues with precedence and order of operations if parameters are not properly quoted and an expression was passed.
```cpp
#define TWO 1+1

TWO * TWO; // 1+1 * 1+1 = 1 + (1 * 1) + 1 = 3
```

Reply:
As for the structured binding bit. If you're not using one of the extracted values, it might not have been appropriate to use structured bindings in the first place. If you only wanted a single value, then accessing that single field makes more sense. Though yes, I get that the syntax is kind of nice, and I do like being able to use it.

Reply:
Those are strawman arguments.

Yes, the win32 header's `max` define is bad. Yes, poorly scoped macros are bad. Yes, you can write macros that abuse the type system. But those are simply examples of _bad code_. That doesn't mean the concept of macros is evil. It's possible to write bad code in any language.

Writing code to generate code is a well-established practice, and I simply don't believe that macros are categorically worse than any other means of doing so. You can write python scripts that generate headers. You can define macros to expand .inc files into a myriad of utility functions. In some cases you can use templates to accomplish your goals. Each way of generating code has its own strengths and weaknesses, and I'd generally argue that you should use the right tool for the right job. That said, there is value in using few tools in general, since each tool has a learning curve. But that's a very weak reason not to use macros.

This is turning quite philosophical.

Reply:
There isn't really a concept of poorly scoped macros. Macros just aren't scoped. To control scope of macros you mostly just isolate them at the translation unit level. Unless perhaps you want to undefine them after use, though that's often not very practical. The next best alternative is to give the name some kind of lengthy unique prefix, which makes them less pleasant to use.

Yes, you can write bad code in any language. Though with that said, some features are more error prone than others. Macros are a particularly error prone feature, and the language has evolved in a way that there are now often alternatives to using them.

The point is more that when there are alternatives that don't involve using macros, then alternative solutions should perhaps be used. Inline functions are a better alternative to function macros. Template functions are a better alternative to untyped macro functions that need to work with multiple types. Typed constants are a better alternative to untyped macro defined values.

If your use is X-Macros, then yes, you do probably need to use macros. There's no real alternative to that offered by other language features.

Similarly, template functions provide a rather limited and controlled way of doing code gen, which doesn't include every possible use case of macros. For instance, you can't declare a variable inside an enclosing scope by calling a template. You'd need a macro to accomplish that. Though that's something people often go to great lengths to avoid accidentally doing when writing macro. Adding a scope block to a macro is uglier than many people realize. See:
https://stackoverflow.com/questions/1067226/c-multi-line-macro-do-while0-vs-scope-block

As for the whole strawman thing, you did literally ask "what's wrong with macros?", and there are plenty of problems that come up when using them. And there's more that haven't been thought of here. For instance, the coming module system for C++ has complications due to macros.

----

In terms of alternatives for the unused variable warning, I generally see them in two places:

```cpp
void someFunction(int unused); // Warning
void someFunction(int /*unused*/); // No warning
```

```cpp
auto [unused, b] = someStructure; // Warning
auto b = someStructure.b; // No warning
```

Reply:
>The point is more that when there are alternatives that don't involve using macros, then alternative solutions should perhaps be used. Inline functions are a better alternative to function macros. Template functions are a better alternative to untyped macro functions that need to work with multiple types. Typed constants are a better alternative to untyped macro defined values.

>If your use is X-Macros, then yes, you do probably need to use macros. There's no real alternative to that offered by other language features.

As often is the case, I think our views are actually not that different now that we've elaborated a bit. I mistook your initial feedback for "you should never use macros," but re-reading that part I see you never wrote anything to that effect.

> As for the whole strawman thing, you did literally ask "what's wrong with macros?"

You're right - I should have asked "what's wrong with this macro?" :-) But let's not go into that discussion.